# WGCNA_of_T1DM_associated_TBI
This repository contains code scripts and output files for the project "Identification of key genes and enriched biological processes in T1DM-associated traumatic brain injury through weighted gene correlation network and bioinformatics analysis."
T1DM can coexist with TBI and recent studies have indicated that diabetes (T1DM) exacerbates the outcomes of Traumatic brain injury (TBI), leading to more severe cognitive deficits and increased risk of complications. This study sought to investigate the underlying molecular mechanisms and potential therapeutic targets for T1DM-associated TBI.

Four microarray datasets related to Type 1 diabetes (T1DM) and Traumatic brain injury (TBI) downloaded from GEO repository were used in this study. The datasets were normalized, filtered, and batch corrected using TMM-normalization method in "edgeR" package and "ComBat" function in "sva" package. The processed data were analyzed using "Limma" bioconductor package to screen for differentially expressed genes (DEGs). Using whole expression data, WGCNA analysis was performed to identify significant genes in disease-related modules using hierarchical clustering of genes and "Spearman correlation". GSEA was used to identify enriched biological processes of DEGs. GO and KEGG analysis using "clusterProfiler" were used for functional enrichment annotation of the module genes. Protein Network analysis was constructed using "STRING" and the network was exported into "Cytoscape". Next, the "cytoHubba" plugin was used for hub genes identification. Genes that are significantly associated with both T1DM and TBI were identified based on the overlap of TIDM-related module genes and TBI-related module genes. Finally, the identified shared genes were functionally annotated using "Metascape" and "Reactome" to further understand their biological functions in T1DM-associated TBI.

A total of 284 DEGs were identified in T1DM, of which 11 were upregulated and 9 were downregulated. GSEA analysis showed that the DEGs were significantly enriched in "lipid metabolic process and response to xenobiotic stimulus". A total of 584 DEGs were identified in TBI, of which 186 were upregulated and 9 were downregulated. GSEA analysis showed that the DEGs were mainly enriched in "immune response-regulating signaling pathway". Using WGCNA, 122 genes were identified in T1DM-related modules, and 368 genes were identified in TBI-related modules. GO and KEGG analysis showed that T1DM-related module genes were correlated with "lipid metabolic process and stress response", and the TBI-related module genes were correlated with "adaptive immune response". 

PPI network analysis identified 20 hub genes (...,...,...,..) from T1DM-related module genes. The hub genes were primarily related to "ribosome biogenesis and RNA post-transcriptional regulation". PPI network analysis identified 20 hub genes (....,...,....) from TBI-related module genes. The hub genes were primarily related to "inflammatory mediators and immune response". 

Based on the overlap of T1DM-related module genes and TBI-related module genes, ...., ...., and .... were identified as key genes of T1DM-associated TBI. Functional enrichment annotation using Metascape showed that the shared genes were primarily enriched in cellular response to lipid hydroperoxide, cytokine mediated receptor signaling activity, regulation of sequestering of triglyceride, negative regulation of IL-12 production, positive regulation of lipid storage, and positive regulation of macrophage chemotaxis. Reactome showed that they were primarily related to inflammation, neutrophil degranulation, and lipid storage, respectively. Thus, we concluded that dysregulation of lipid sequestering and neuroinflammation are the potential molecular mechanisms of T1DM-associated TBI. ..., ..., and .... are positively correlated with T1DM-associated TBI, and may be important biomarkers and potential treatment targets for diabetic TBI.
